src/Avalonia.Controls/Control.cs

```diff
@@ -215,6 +273,49 @@ void IVisualBrushInitialize.EnsureInitialized()
         /// <returns>The control that receives the focus adorner.</returns>
         protected virtual IControl? GetTemplateFocusTarget() => this;
 
+        /// <summary>
+        /// Invoked as the first step of marking the control as loaded and raising the
+        /// <see cref="Loaded"/> event.
+        /// </summary>
+        internal void OnLoadedCore()
```

Comment:
I need someone more familiar with Avalonia's rendering system to let me know where to call this. Ideally it is called just before or even just after the control is first rendered.

WPF: https://docs.microsoft.com/en-us/dotnet/api/system.windows.frameworkelement.loaded?view=windowsdesktop-6.0
UWP: https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.frameworkelement.loaded?view=winrt-22621

Reply:
Also note that @amwx suggestion was some place in `LayoutManager.ExecuteLayoutPass`

Reply:
Just for case: WPF keeps track of elements that haven't loaded/unloaded and need to at various points prior to render. It's quite difficult to track this since the code jumps around (some of it's in FrameworkElement), but there's quite a bit of managing the subtree to each FrameworkElement, ensuring Loaded/Unloaded events don't conflict with each other. They also have a separate DispatcherPriority.Loaded, but there's a comment somewhere that not all paths use the dispatcher for Loaded/Unloaded.

When the render message is received, it fires all those events: 
`MediaContext.RenderMessageHandlerCore` -> `MediaContext.FireInvokeOnRenderCallbacks` -> `MediaContext.FireLoadedPendingCallbacks`

https://github.com/dotnet/wpf/blob/42a8bbe349253e71d53e0d24ec18363dcb20655d/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Media/MediaContext.cs#L1841-L1857
https://github.com/dotnet/wpf/blob/42a8bbe349253e71d53e0d24ec18363dcb20655d/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Media/MediaContext.cs#L1950-L2014

Reply:
Each control has an IsLoaded property in this PR. That tracks whether the control itself is loaded and I don't think you need a separate store to track everything. The only thing that may be tricky is ensuring that all the children and template controls are created before calling OnLoadedCore. For someone that knows the layout and rendering system in detail I have to think there is a place where these conditions are met and OnLoadedCore should be called.

Reply:
I am not sure what's the best solution here, and it probably will be changed with Composition API PR.
But at the very least I see possibility to schedule a "LoadedCore" call on Dispatcher with Render priority. And this scheduling can happen inside of OnAttachedToVisualTree, since it's expected to always be called before Loaded event.

There should be additional checks in OnLoadedCore just in case if control was detached in between scheduling and first render.

Reply:
This plain simple code might also work, but adds small condition to each Render execution
```csharp
override void Render()
{
    if (!_isLoaded) OnLoadedCore();
    base.Render();
}
```

Reply:
@kekekeks @grokys

Reply:
Doing this async somehow is probably a good idea too if we are tapping into the Render() calls. I'm not sure the effects to rendering if controls do something slow in the Loaded event.

Reply:
There shouldn't be much difference between slowing down Render call or OnAttached call, both are executed on UI thread. Drawing context inside of Render call will defer draw commands to the render thread.
It's different for immediate renderer though.

Reply:
OnAttachedToVisualTree happens way before layout. IIRC WPF's `Loaded` event happens after layout.

Reply:
> OnAttachedToVisualTree happens way before layout.

Well, if is scheduled as Render priority on the dispatcher as @maxkatz6 said I also think it should work to use this to start it. However, I like the Render() override better -- cleaner, easier to manage.

> IIRC WPF's Loaded event happens after layout.

I think it would be functionally equivalent in Avalonia to also trigger this just before render. In practice it should be the same.

> This plain simple code might also work, but adds small condition to each Render execution

I like this idea for now and will add it shortly.

Reply:
Tapping to `Render` will skip layout for any changes made by the `Loaded` handler. So you'll have a frame with the old content visible.

Reply:
I'll read more in the existing docs for WPF/UWP. Generally, I would expect `Loaded` to occur after layout and everything is done. If an app does significant changes in the `Loaded` event I think a flicker is expected as things update in the next layout/render pass. The `Loading` event was useful for that case but Avalonia already has OnAddedToVisualTree so I'm not adding the `Loading` event along with `Loaded` and `Unloaded` at this time.

Reply:
WPF has a separate priority for `Loaded` event. They schedule it _after_ the first render.

https://docs.microsoft.com/en-us/dotnet/api/system.windows.threading.dispatcherpriority?view=windowsdesktop-6.0

>Loaded | 6 | The enumeration value is 6. Operations are processed when  layout and render has finished but just before items at input priority  are serviced. Specifically this is used when raising the Loaded event.

Reply:
This could be duplicated without a new priority i think. Just schedule raising the loading event (with render priority) while doing the first render. It will be raised once render is complete. Either way though, I think the ideas discussed here are already a close-enough approximation to be usable in almost all scenarios.

Reply:
So... apparently Avalonia already has the Loaded priority in the Dispatcher. That makes this a whole lot easier and I'm just going to schedule it in the `OnAttchedToVisualTreeCore` method. This actually makes the implementation fairly clean and matches with the Unloaded strategy (called from `OnDetachedFromVisualTreeCore`).

I'm also assuming Loaded priority is fully implemented as it is used in two places already.

Reply:
I'd suggest to add the visual to a global HashSet rather than directly using Dispatcher and only do that if there are `Loaded` event subscribers. That should save quite a few allocations.

Reply:
> I'd suggest to add the visual to a global HashSet

Could you explain in more detail this statement? It isn't clear how to actually raise the event doing this or where the set is managed. Do you mean tap into the internal dispatcher code directly?

Reply:
I think he means something like we have for weak events:
- shared hashset (probably doesn't need to be 'weak' in this PR): https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Utilities/WeakEvent.cs#L105
- when any control needs to call OnLoaded, it will be added to the hashset of scheduled controls
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Utilities/WeakEvent.cs#L126
- and call Dispather once if it wasn't already called
- inside of the Dispather callback read all the controls to be loaded and clear hashset 

I can image it to be done with some kind of internal static manager.

Reply:
I really hope all that isn't required for this PR. Especially before there is any confirmed performance issue. Not to mention the queueing of events might not be ideal for some apps. (Better to process incrementally in some cases than all at once)

That said, the leak tests seem to be failing now for some reason.

Reply:
Well, with no further comments I will just make the changes to queue Loaded events. Getting this PR done is more important than debating various implementations at this point.

Reply:
I guess one thing I don't understand about using the Loaded dispatcher priority is how it handles multiple controls. It seems to me that the dispatcher would fire based on global state and not the state of a specific control. Since Avalonia doesn't have x:Load like UWP I guess this isn't an issue for now. However, in the future, it may be necessary again to handle each control separately as they will have different lifecycles and be loaded at different times. The global Loaded dispatcher priority wouldn't work there.

I'm also not a fan of using HashSet in this case but will follow the guidance. Finally, AFAIK it isn't possible to check for subscribers with routed events.

Otherwise, please take another look and review.

Reply:
Solution with hashset helps us to avoid multiple dispatcher jobs, which isn't terribly slow, but slower that simple hashset.

Reply:
Marking this conversation as resolved since the design is largely decided now.